### ISSUE TITLE
OpenGD77 channels power

### ISSUE THREAD

**Issue by ua0lnj** [hmatuschek/qdmr#507]
Hi.
OpenGD77 have "Master" settings in channels power option. Master-min-low-midle-hi-max. After first codeplug upload from Qdmr I have min power of all channels now. Master mast set as default, if power not setting in channels options.
That's great, do everything from Linux, no need to use Win :)

**Comment by hmatuschek** [comment]
First of all, thanks for reporting. It is certainly a bug. Unfortunately, I don't know, where the global power setting is stored and I don't have a OpenGD77 radio at hand. But you can help. 

1. Read the binary codeplug from the radio. `dmrconf read codeplug1.dfu`
2. Change the global power setting in the radio
3. Read it again `dmrconf read codeplug2.dfu`
4. For each DFU file, create a hex-dump: `dmrconf info codeplugX.dfu > codeplugX.hex` 
5. Compare them using diff. E.g., `diff codeplug1.hex codeplug2.hex` 

If the global power setting is stored in the codeplug on the radio, the difference should be visible.

**Comment by hmatuschek** [comment]
Also, I have fixed the power decoding of the channel power setting 
`507-opengd77-channels-power` unfortunately, you may need to reset all power settings of your channels.

**Comment by ua0lnj** [comment]
Ok, I'll try tomorrow.

**Comment by ua0lnj** [comment]
I did it... 0 size of diff.
Another case, change the settings of one channel and compare?

**Comment by ua0lnj** [comment]
About 507-opengd77-channels-power.
I read and wrote back codeplug. All channels were master power, became 500 mW. In Qdmr were default power, became low power, and no default.

**Comment by hmatuschek** [comment]
Yes, that was a bug in de- and then reencoding of the channels. You need to
reset all channels to "default", because they are now all explicitly set to
500mw.

Andrey ***@***.***> schrieb am Mi., 25. Dez. 2024, 14:59:

**Comment by hmatuschek** [comment]
Concerning the 0-diff. This means, the FW stores these settings somewhere
else. Those global settings should be stored in the codeplug. Maybe, we
should suggest that to the OpenGD77 guys.

Andrey ***@***.***> schrieb am Mi., 25. Dez. 2024, 14:19:

**Comment by ua0lnj** [comment]
Global power settings not stored in codeplug. I change master power, restored settings -> download sawed codeplug (In Windows OpenGD CPS), and master power not changed.

**Comment by ua0lnj** [comment]
It clearly doesn't work correctly. I set default power on the channel, write to the radio, read from the radio and get low but not default. It seems that the channel power is read correctly, but is written incorrectly. If I change the channel settings in the radio, the program read correctly.

**Comment by hmatuschek** [comment]
> Qdmr does not read global power settings correctly. I change it, but Qdmr read High always.

Since the global power setting is not stored by the firmware inside the codeplug (for some unknown reason), I cannot read it. So, I default to "high". But since it is not stored, I cannot set it too. Therefore, this setting has no effect here. 

I consider this a bug in OpenGD77, I would make much more sense, to store the global power setting within the codeplug, if it is referenced (inside the channel settings) there.

**Comment by hmatuschek** [comment]
> It clearly doesn't work correctly. I set default power on the channel, write to the radio, read from the radio and get low but not default. It seems that the channel power is read correctly, but is written incorrectly. If I change the channel settings in the radio, the program read correctly.

Thanks for this observation, it helps a lot.

**Comment by hmatuschek** [comment]
> It clearly doesn't work correctly. I set default power on the channel, write to the radio, read from the radio and get low but not default. It seems that the channel power is read correctly, but is written incorrectly. If I change the channel settings in the radio, the program read correctly.

Added a unit test for this, qdmr appears to be self-consistent. That is, it is written such that the default power is read by qdmr correctly. This means, that the encoding of this setting is more complicated and the firmware interferes with the settings written to the device.

**Comment by hmatuschek** [comment]
Okay, can reproduce it in unit tests. It has nothing todo with the OpenGD77 codeplug but with the preprocessing of the codeplug.

**Comment by ua0lnj** [comment]
Yes, it's works now!!!
